src/main/java/ch/njol/skript/effects/EffToggle.java

```diff
@@ -40,63 +41,84 @@
+			"(toggle|switch) [[the] state of] %booleans/blocks%"
+			);
+	}
+
+	private enum State {
+    ACTIVATE, DEACTIVATE, TOGGLE
 	}
 
 	@SuppressWarnings("null")
-	private Expression<Block> blocks;
-	private int toggle;
+	private Expression<?> toggledExpr;
+	private State state;
 	
-	@SuppressWarnings({"unchecked", "null"})
 	@Override
-	public boolean init(final Expression<?>[] vars, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
-		blocks = (Expression<Block>) vars[0];
-		toggle = matchedPattern - 1;
+	public boolean init(Expression<?>[] vars, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		toggledExpr = (Expression<?>) vars[0];
+		state = State.values()[matchedPattern];
+		if (toggledExpr.getReturnType() == Boolean.class && !ChangerUtils.acceptsChange(toggledExpr, ChangeMode.SET, Boolean.class)) {
+			Skript.error(toggledExpr.toString(null, false) + " cannot be toggled");
+			return false;
+		}
 		return true;
 	}
 
 	@Override
-	protected void execute(final Event e) {
-		for (Block b : blocks.getArray(e)) {
-			BlockData data = b.getBlockData();
-			if (toggle == -1) {
-				if (data instanceof Openable)
-					((Openable) data).setOpen(false);
-				else if (data instanceof Powerable)
-					((Powerable) data).setPowered(false);
-			} else if (toggle == 1) {
-				if (data instanceof Openable)
-					((Openable) data).setOpen(true);
-				else if (data instanceof Powerable)
-					((Powerable) data).setPowered(true);
-			} else {
-				if (data instanceof Openable) // open = NOT was open
-					((Openable) data).setOpen(!((Openable) data).isOpen());
-				else if (data instanceof Powerable) // power = NOT power
-					((Powerable) data).setPowered(!((Powerable) data).isPowered());
+	protected void execute(Event event) {
+		ArrayList<Object> toggledValues = new ArrayList<>();
+		for (Object obj : toggledExpr.getArray(event)) {
+			if (obj instanceof Block) {
+				Block block = (Block) obj;
+				BlockData data = block.getBlockData();
+				if (state == State.TOGGLE) {
+					if (data instanceof Openable) // open = NOT was open
+						((Openable) data).setOpen(!((Openable) data).isOpen());
+					else if (data instanceof Powerable) // power = NOT power
+						((Powerable) data).setPowered(!((Powerable) data).isPowered());
+				} else {
+					boolean value = state == State.ACTIVATE; 
+					if (data instanceof Openable)
+						((Openable) data).setOpen(value);
+					else if (data instanceof Powerable)
+						((Powerable) data).setPowered(value);
+				}
+
+				block.setBlockData(data);
+				toggledValues.add(block);
+
+			} else if (obj instanceof Boolean) {
+				toggledValues.add(!(Boolean) obj);
 			}
-			
-			b.setBlockData(data);
 		}
-	}
 
+		List<Object> filteredValues = toggledValues.stream().filter(
+			obj -> ChangerUtils.acceptsChange(toggledExpr, ChangeMode.SET, obj.getClass())
+		).collect(Collectors.toList());
+		toggledExpr.change(event, filteredValues.toArray(), ChangeMode.SET);
```

Comment:
We might need to note a behavior change here. For changing something like a list, I believe the indexes will now be lost, which was not prior behaviors. We *might* choose to only rewrite the list if it contains a boolean?

Reply:
When you're talking about the list, are you talking about the ``toggledExpr`` or the ``filteredValues``, or even both?

Reply:
if we have a list like `{_values::*}` with:
`{_values::pickle::name} = "APickledWalrus"`
`{_values::pickle::likesCake} = true`
if we then do `toggle {_values::*}`, the indices such as `name` and `likesCake` will be lost (replaced by `1` and `2`).
This is probably unavoidable. However, if we have a list like
`{_values::pickle::name} = "APickledWalrus"`
`{_values::pickle::githubName} = "APickledWalrus"`
if we then do `toggle {_values::*}`, there are no booleans to toggle, meaning we should not update the list (and cause it to lose its indices)

Reply:
Looking at the code, don't i already do it?

https://github.com/SkriptLang/Skript/pull/4154/files#diff-4cdf11cdcad1ff81963a3380536868678c5781162395d093ecd42b33c71add01R74

Also, tests are failing, did i misunderstand what you meant?

Reply:
There may have been some confusion, sorry.

Given a list containing` "one" and "two"`, doing `toggle {_list::*}` should not modify the list since there is nothing to toggle (meaning the indexes will be preserved).

However, given a list containing `"one", "two" and true`, doing `toggle {_list::*}` should modify the list since the third value is a boolean (meaning the indexes will be lost).

I think the fix you made is correct, but the test is wrong.

Working with the testing you added, you should be testing:
```
set {_list::myFirstIndex::myFirstValue} to "hello"
set {_list::mySecondIndex::mySecondValue} to "world"
toggle {_list::*}
assert {_list::myFirstIndex::myFirstValue} is "hello" with "{_list::myFirstIndex::myFirstValue} should be toggled to 'hello'"
assert {_list::mySecondIndex::mySecondValue} is "world" with "{_list::mySecondIndex::mySecondValue} should be toggled to 'world'"
```

Reply:
But the test already test all of that, right? I mean, the only difference here is there is not any boolean index tested, but users could have a boolean index, so it becomes important to test it, don't you think?

Reply:
Update: shouldn't we throw an error instead? If the user tries to toggle strings, it shouldn't work. If the variable list includes string, but also boolean, shouldn't we throw an error? Note that i don't think we should throw an error if all values are togglables.

Reply:
Well, we can't really do runtime errors. I don't think it is just an issue to ignore the values and only change the ones that are actually togglable. You could also opt to not make *any* changes if there is an invalid type.

The main point still stands, though. If the list contains only blocks, we do not want to actually call the change method as it will reset the indices.

Reply:
it looks reasonable
you may want to check out https://github.com/SkriptLang/Skript/pull/7120's changeInPlace utility method

Reply:
I would need a usage example please. That being said, i will wait for this PR to be merged before doing any change regarding this change. 

In terms of organization, either the ``changeInPlace`` PR is merged before this one, and so i will apply the change here, or this PR is merged before, and so i will apply the change in the other PR.

In any case, we need to merge one of them to move on the other.